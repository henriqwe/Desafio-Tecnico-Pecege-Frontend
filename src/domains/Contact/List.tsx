import { IContact } from '&operations/queries/fetchContacts'
import { Table } from '&components/Table/Table'
import { Input } from '&components/Input/Input'

import { useForm } from 'react-hook-form'
import { useEffect, useState } from 'react'
import { sortArrayByKey } from '&utils/handles/sortArrayByKey'
import { useContacts } from '&contexts/contactsContext'
import { useNavigate } from 'react-router-dom'
import { Loading } from '&components/Loading/Loading'
import { routes } from '&utils/routes'

export function ContactList() {
  const navigate = useNavigate()
  const { contactsQuery, removeContact } = useContacts()

  const [contacts, setContacts] = useState<IContact[]>()
  const [ascendentName, setAscendentName] = useState(true)

  const { control } = useForm()

  function handleData(data: IContact[], asc: boolean) {
    setContacts(sortArrayByKey({ data, asc, key: 'name' }))
  }

  function handleOnChange(value: string) {
    const newDataList = contactsQuery?.data?.filter(
      (contact) =>
        contact.name.toLocaleLowerCase()?.includes(value.toLocaleLowerCase())
    )
    handleData(newDataList as IContact[], ascendentName)
  }

  function toggleAscendentName(data: IContact[]) {
    handleData(data, !ascendentName)
    setAscendentName(!ascendentName)
  }

  useEffect(() => {
    if (contactsQuery?.data) {
      handleData(contactsQuery?.data, ascendentName)
    }
  }, [contactsQuery?.data])

  if (contactsQuery?.isLoading) {
    return <Loading />
  }

  return (
    <div>
      <h1>Desafio TÃ©cnico Pecege - Frontend</h1>

      <Input
        control={control}
        name="name"
        handleChangeDebounce={handleOnChange}
        label="Nome"
      />
      <div onClick={() => navigate(routes.createContact.path)}>ADICIONAR</div>
      <Table
        data={contacts}
        titles={[
          {
            name: 'Nome',
            key: 'name',
            onClick: {
              fn: () => toggleAscendentName(contacts as IContact[]),
              ascIcon: ascendentName
            }
          },
          { name: 'Telefone', key: 'phone' },
          { name: 'Email', key: 'email' }
        ]}
        actions={(contact: IContact) => (
          <td>
            <div
              onClick={() =>
                navigate(routes.editContact.path(contact.id.toString()))
              }
            >
              EDITAR
            </div>
            <div onClick={() => removeContact(contact.id)}>REMOVER</div>
          </td>
        )}
      />
    </div>
  )
}
