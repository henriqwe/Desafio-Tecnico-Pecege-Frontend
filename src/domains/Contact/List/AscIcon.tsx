import { BarsArrowDownIcon, BarsArrowUpIcon } from '@heroicons/react/20/solid'

export function AscIcon({ asc }: { asc: boolean }) {
  if (asc) {
    return <BarsArrowUpIcon style={{ width: '1.3rem', height: '1.3rem' }} />
  }
  return <BarsArrowDownIcon style={{ width: '1.3rem', height: '1.3rem' }} />
}
