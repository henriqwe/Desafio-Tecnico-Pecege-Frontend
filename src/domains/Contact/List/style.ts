import styled from 'styled-components'
import * as Icons from '@heroicons/react/24/outline'

export const CardList = styled.div`
  display: flex;
  flex-direction: column;
  gap: 0.8rem;
  overflow-x: auto;
  flex: 1;
  padding-right: 0.8rem;
  max-height: calc(100vh - 11rem);
  min-height: calc(100vh - 11rem);
`

export const Container = styled.section`
  position: relative;
  padding: 0.5rem;
  max-height: 100vh;
  min-height: 100vh;
  display: flex;
  flex: 1;
  flex-direction: column;
  overflow-x: none;
`
export const Separator = styled.div`
  width: 100%;
  border: 1px solid rgba(68, 67, 90, 0.3);
  margin-bottom: 1rem;
`
export const ListHeading = styled.div`
  display: flex;
  justify-content: space-between;
  align-items: end;
  flex: 1;
  width: 100%;
`
export const OrderByNameContent = styled.div`
  display: flex;
  align-items: end;
  gap: 0.3rem;
  color: rgb(30 41 59);
`

export const UserPlusIcon = styled(Icons.UserPlusIcon)`
  width: 1.3rem;
  height: 1.3rem;
  color: rgb(15 23 42);
  font-weight: 600;
`
export const ActionsContainer = styled.div`
  display: flex;
  align-items: center;
  justify-content: space-between;
  gap: 1rem;
  margin-top: 1rem;
  width: 100%;
`
export const ButtonText = styled.span`
  font-weight: 700;
`
export const IllustrationContainer = styled.div`
  display: flex;
  flex: 1;
  flex-direction: column;
  gap: 3rem;
  align-items: center;
  padding-top: 6rem;
`
export const LoadingContainer = styled.div`
  display: flex;
  flex: 1;
  flex-direction: column;
  gap: 3rem;
  align-items: center;
  padding-top: 6rem;
`
export const IllustrationWrapper = styled.div`
  height: 12rem;
  width: 12rem;
`
export const IllustrationMessage = styled.span`
  font-size: 1.1rem;
`
export const BarsArrowDownIcon = styled(Icons.BarsArrowDownIcon)`
  width: 1.3rem;
  height: 1.3rem;
`
export const BarsArrowUpIcon = styled(Icons.BarsArrowUpIcon)`
  width: 1.3rem;
  height: 1.3rem;
`
