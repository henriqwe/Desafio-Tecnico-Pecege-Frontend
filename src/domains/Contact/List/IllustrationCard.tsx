import { DreamerIllustration } from '&components/Illustrations/DreamerIllustration/DreamerIllustration'
import { PeopleSearchIllustration } from '&components/Illustrations/PeopleSearchIllustration/PeopleSearchIllustration'
import * as S from '&domains/Contact/List/style'

export function IllustrationCard({
  isEmpty,
  notFound
}: {
  notFound: boolean
  isEmpty: boolean
}) {
  if (!isEmpty && !notFound) {
    return <></>
  }
  const illustration = isEmpty ? (
    <DreamerIllustration />
  ) : (
    <PeopleSearchIllustration />
  )

  const message = isEmpty
    ? 'Lista de contatos está vazia'
    : 'Contato não encontrado'

  return (
    <S.IllustrationContainer>
      <S.IllustrationWrapper>{illustration}</S.IllustrationWrapper>
      <S.IllustrationMessage>{message}</S.IllustrationMessage>
    </S.IllustrationContainer>
  )
}
