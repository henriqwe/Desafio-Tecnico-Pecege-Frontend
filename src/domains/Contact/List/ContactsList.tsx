import type { IContact } from '&types/contact'
import { fetchContacts } from '&services/queries/fetchContacts'
import { useEffect, useState } from 'react'
import { sortArrayByKey } from '&utils/handles/sortArrayByKey'
import { useNavigate } from 'react-router-dom'
import { routes } from '&utils/routes'
import { ContactCard } from '&components/ContactCard/ContactCard'
import * as S from '&domains/Contact/List/style'
import { IllustrationCard } from '&domains/Contact/List/IllustrationCard'
import { useQuery, useQueryClient } from '@tanstack/react-query'
import { FloatingActionButton } from '&components/FloatingActionButton/FloatingActionButton'

interface props {
  ascendentName: boolean
  nameValue?: string
}

export function ContactList({ ascendentName, nameValue }: props) {
  const navigate = useNavigate()
  const [contactsFiltered, setContactsFiltered] = useState<IContact[]>()
  const queryClient = useQueryClient()

  async function handleFetch() {
    const contacts = queryClient.getQueryData(['contacts']) as IContact[]

    if (contacts) {
      return contacts
    }
    return fetchContacts()
  }
  const contactsQuery = useQuery({
    queryKey: ['contacts'],
    queryFn: () => handleFetch(),
    refetchOnWindowFocus: false,
    retry: false,
    suspense: true
  })

  const showIllustration =
    (contactsFiltered?.length === 0 ||
      contactsQuery?.data?.length === 0 ||
      contactsQuery?.data === undefined) &&
    !contactsQuery?.isLoading

  function onFilter(text?: string) {
    if (!text) {
      return contactsQuery?.data as IContact[]
    }
    const newDataList = contactsQuery?.data?.filter((contact) =>
      contact.name.toLocaleLowerCase()?.includes(text.toLocaleLowerCase())
    )
    return newDataList as IContact[]
  }

  function handleData(data: IContact[], asc: boolean) {
    setContactsFiltered(sortArrayByKey({ data, asc, key: 'name' }))
  }

  function handleNavigate(id: number) {
    navigate(routes.viewContact.path(id.toString()))
  }

  useEffect(() => {
    if (contactsQuery?.data) {
      const data = onFilter(nameValue)
      handleData(data, ascendentName)
    }
  }, [contactsQuery?.data, ascendentName, nameValue])

  return (
    <S.CardList>
      {showIllustration ? (
        <IllustrationCard
          notFound={contactsFiltered?.length === 0}
          isEmpty={
            contactsQuery?.data?.length === 0 ||
            contactsQuery?.data === undefined
          }
        />
      ) : (
        contactsFiltered?.map((contact) => (
          <ContactCard
            key={contact.phone + contact.id}
            id={contact.id}
            phone={contact.phone}
            name={contact.name}
            email={contact.email}
            navigateAction={handleNavigate}
          />
        ))
      )}

      <FloatingActionButton />
    </S.CardList>
  )
}
