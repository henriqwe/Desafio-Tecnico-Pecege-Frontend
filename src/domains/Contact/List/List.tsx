import { Input } from '&components/Input/Input'
import { useForm } from 'react-hook-form'
import { Suspense, useState } from 'react'
import { useNavigate } from 'react-router-dom'
import { routes } from '&utils/routes'
import * as S from '&domains/Contact/List/style'
import { Button } from '&components/Button/Button'
import { ContactList } from './ContactsList'
import { SkeletonTheme } from 'react-loading-skeleton'
import { ContactCardSkelton } from '&components/ContactCard/ContactCardSkeleton'

export function List() {
  const navigate = useNavigate()

  const [ascendentName, setAscendentName] = useState(true)
  const { control } = useForm()
  const [nameValue, setNameValue] = useState<string>()

  function toggleAscendentName() {
    setAscendentName(!ascendentName)
  }

  function handleOnChange(text: string) {
    setNameValue(text)
  }

  return (
    <S.Container>
      <S.Header>
        <h1>Home</h1>
      </S.Header>

      <S.ActionsContainer>
        <S.InputWrapper>
          <Input
            control={control}
            name="name"
            handleChangeDebounce={handleOnChange}
            label="Buscar"
            debounceDelay={1000}
            data-testid={'search-input'}
          />
        </S.InputWrapper>
        <div>
          <Button
            onClick={() => navigate(routes.createContact.path)}
            variant="success"
            data-testid={'button-new-contact'}
          >
            <S.ButtonText>Novo</S.ButtonText> <S.UserPlusIcon />
          </Button>
        </div>
      </S.ActionsContainer>
      <S.ListHeading>
        <S.Title>Contatos</S.Title>

        <S.OrderByNameContent
          onClick={() => toggleAscendentName()}
          data-testid={'button-sort-name'}
        >
          <span>Nome</span>
          {ascendentName ? <S.BarsArrowUpIcon /> : <S.BarsArrowDownIcon />}
        </S.OrderByNameContent>
      </S.ListHeading>
      <S.Separator />

      <S.CardList>
        <Suspense
          fallback={
            <S.CardList>
              <SkeletonTheme baseColor="#d9d8df" highlightColor="#bebbbb">
                {Array.from({ length: 9 }).map((_, idx) => (
                  <ContactCardSkelton key={'Skeleton' + idx} />
                ))}
              </SkeletonTheme>
            </S.CardList>
          }
        >
          <ContactList ascendentName={ascendentName} nameValue={nameValue} />
        </Suspense>
      </S.CardList>
    </S.Container>
  )
}
