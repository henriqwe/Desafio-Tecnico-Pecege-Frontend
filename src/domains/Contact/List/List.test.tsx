import { fireEvent, render, waitFor } from '@testing-library/react'
import '@testing-library/jest-dom'
import { List } from './List'
import { TestProviders } from '&utils/tests/TestProviders'

const mockedUsedNavigate = jest.fn()

jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  useNavigate: () => mockedUsedNavigate
}))
describe('List', () => {
  it('should render the contact list correctly', async () => {
    const { container } = render(
      <TestProviders>
        <List />
      </TestProviders>
    )

    expect(container).toMatchSnapshot()
  })

  it('should navigate to map page', async () => {
    const { findByTestId } = render(
      <TestProviders>
        <List />
      </TestProviders>
    )
    const floatingactionbuttonContainer = await findByTestId(
      'floatingactionbutton-container'
    )

    await waitFor(() => fireEvent.click(floatingactionbuttonContainer))

    expect(mockedUsedNavigate).toBeCalledWith('/mapa')
  })
  it('should navigate to create contact page', async () => {
    const { findByTestId } = render(
      <TestProviders>
        <List />
      </TestProviders>
    )
    const buttonNewContact = await findByTestId('button-new-contact')

    await waitFor(() => fireEvent.click(buttonNewContact))

    expect(mockedUsedNavigate).toBeCalledWith('/cadastrar')
  })

  it('should sort contact name in descendent format', async () => {
    const { findByTestId, findAllByTestId } = render(
      <TestProviders>
        <List />
      </TestProviders>
    )

    const namesAsc = await findAllByTestId('name')
    const phonesAsc = await findAllByTestId('phone')
    const emailsAsc = await findAllByTestId('email')

    expect(namesAsc[0].textContent).toEqual('Chelsey Dietrich')
    expect(phonesAsc[0].textContent).toEqual('(254)954-1289')
    expect(emailsAsc[0].textContent).toEqual('Lucio_Hettinger@annie.ca')

    const button = await findByTestId('button-sort-name')
    await waitFor(() => fireEvent.click(button))

    const namesDesc = await findAllByTestId('name')
    const phonesDesc = await findAllByTestId('phone')
    const emailsDesc = await findAllByTestId('email')

    expect(namesDesc[0].textContent).toEqual('Patricia Lebsack')
    expect(phonesDesc[0].textContent).toEqual('493-170-9623 x156')
    expect(emailsDesc[0].textContent).toEqual('Julianne.OConner@kory.org')
  })

  it('should show a message and illustration in the screen when not found a searched contact', async () => {
    const { findByTestId, findAllByTestId, findByText } = render(
      <TestProviders>
        <List />
      </TestProviders>
    )
    const namesAsc = await findAllByTestId('name')
    const phonesAsc = await findAllByTestId('phone')
    const emailsAsc = await findAllByTestId('email')

    expect(namesAsc[0].textContent).toEqual('Chelsey Dietrich')
    expect(phonesAsc[0].textContent).toEqual('(254)954-1289')
    expect(emailsAsc[0].textContent).toEqual('Lucio_Hettinger@annie.ca')

    const input = (await findByTestId('search-input')) as HTMLInputElement
    await waitFor(() =>
      fireEvent.change(input, { target: { value: 'jhon doe bla' } })
    )

    // promise to await debounce
    await waitFor(async () => await findByText('Contato não encontrado'), {
      timeout: 1200
    })

    const message = (await findByText(
      'Contato não encontrado'
    )) as HTMLSpanElement
    const illustration = (await findByTestId(
      'people-search-illustration'
    )) as HTMLOrSVGElement

    expect(illustration).toBeTruthy()
    expect(message).toBeTruthy()
  })
})
