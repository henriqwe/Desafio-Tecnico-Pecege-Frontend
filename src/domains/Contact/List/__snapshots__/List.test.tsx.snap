// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`List should render the contact list correctly 1`] = `
<div>
  <section
    class="sc-gKHVLF frQkfL"
  >
    <div
      class="sc-fEyylQ btrSVL"
    >
      <h1>
        Home
      </h1>
    </div>
    <div
      class="sc-kDnyCx fxaxYa"
    >
      <div
        class="sc-idyqAC hUsnuD"
      >
        <div>
          <div
            class="sc-dkjKgF gJxtqW"
          >
            <input
              autocomplete="off"
              class="sc-hLclGa gWrBgX"
              data-testid="search-input"
              id="name"
              isinvalid="0"
              label="Buscar"
              placeholder=""
              type="text"
              value=""
            />
            <span
              class="sc-eDnVMP fkBieA"
            >
               
            </span>
          </div>
          <label
            class="sc-bcPKhP hSXdPs"
          />
        </div>
      </div>
      <div>
        <button
          class="sc-dIEoRj djA-dzB"
          data-testid="button-new-contact"
          variant="success"
        >
          <div
            class="sc-hHvkSs ikKKqn"
          >
            <span
              class="sc-ipMvLY jfCAo"
            >
              Novo
            </span>
             
            <svg
              aria-hidden="true"
              class="sc-jrkPcm hWpTG"
              fill="none"
              stroke="currentColor"
              stroke-width="1.5"
              viewBox="0 0 24 24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M19 7.5v3m0 0v3m0-3h3m-3 0h-3m-2.25-4.125a3.375 3.375 0 11-6.75 0 3.375 3.375 0 016.75 0zM4 19.235v-.11a6.375 6.375 0 0112.75 0v.109A12.318 12.318 0 0110.374 21c-2.331 0-4.512-.645-6.374-1.766z"
                stroke-linecap="round"
                stroke-linejoin="round"
              />
            </svg>
          </div>
        </button>
      </div>
    </div>
    <div
      class="sc-ftLKQv bhTMLu"
    >
      <h3
        class="sc-bjEwCx evvuaY"
      >
        Contatos
      </h3>
      <div
        class="sc-pqitP bGDaTT"
        data-testid="button-sort-name"
      >
        <span>
          Nome
        </span>
        <svg
          aria-hidden="true"
          class="sc-fnOeiS dEmuwk"
          fill="none"
          stroke="currentColor"
          stroke-width="1.5"
          viewBox="0 0 24 24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M3 4.5h14.25M3 9h9.75M3 13.5h5.25m5.25-.75L17.25 9m0 0L21 12.75M17.25 9v12"
            stroke-linecap="round"
            stroke-linejoin="round"
          />
        </svg>
      </div>
    </div>
    <div
      class="sc-iBAaJG fEItki"
    />
    <div
      class="sc-jSwlEQ kKbemP"
    >
      <div
        class="sc-eLSjS imMixF"
      />
    </div>
  </section>
</div>
`;
